README.md
# Jekyll URL Helper Filter Plugin with CDN Support

This is a Jekyll [Liquid Filter](https://github.com/Shopify/liquid/wiki/Liquid-for-Programmers) plugin that allows you to abstract your URLs in __template__ and __\_posts__ files so that your code is portable and normalized to a greater degree.

By using this plugin you only ever have to change your URL components in one place and the change will be reflected throughout your entire site.

This is highly advantageous with settings like `_config.yml` -> `baseurl`, your `feed.xml` file (which needs absolute URLs) and if you use a CDN or the cloud to serve some of your assets.

__Version__: 1.0.0

## How to Install

To install this plugin [download](https://github.com/jhauraw/jekyll-url-helper-filter-plugin/archive/master.zip) or [copy](https://github.com/jhauraw/jekyll-url-helper-filter-plugin) it into a directory named `_plugins` in your project working directory.

> Visit the official __Jekyll__ [Plugin's Documentation](http://jekyllrb.com/docs/plugins/) for more information.

## Dependencies

This plugin requires the Ruby [Zlib](http://zlib.net/) library, which should be installed on most machines with a recent version of Ruby.

## Usage

See the plugin itself for information on the methods. For a high-scalability implementation and lots of usage ideas check out my own [website repo](https://github.com/jhauraw/jhaurawachsman.com).

### Methods that begin with to\_

The methods which begin with `to_` transform a _root-relative_ path/url into the path/url __type__ of the method's name, e.g., base, absolute or cdn: `to_<type>url`.

The methods can be used on assets such as CSS and JS files, images, XML feeds in template files and anywhere you find useful.

### method: to\_baseurl

Transform a _root-relative_ URL into a _base-relative_ URL.

In your templates and optionally in your Markdown files use the `to_baseurl` method to __append__ the `baseurl` parameter value in `_config.yml` to _root-relative_ URLs.

Example:

In `_config.yml` you have:

    baseurl: /blog

In `_layouts` -> `default.html` you will likely have a CSS file linked, so we can add the _filter_ to the _Liquid_ tag to abstract the URL:

    <link rel="stylesheet" href="{{ '/css/site.css' | to_baseurl }}" />

When you _generate_ your site, the result is:

    <link rel="stylesheet" href="/blog/css/site.css" />

### method: to\_absurl

Transform a _root-relative_ URL into an _absolute_ URL.

In your templates and optionally in your Markdown files use the `to_absurl` method to __append__ the `url` parameter value in `_config.yml` to _root-relative_ URLs.

Example:

In `_config.yml` you have:

    url: http://www.domain.tld

In `feed/index.xml` you will likely have a _self_ link, so we can add the _filter_ in a _Liquid_ tag to make it an _absolute_ and _abstracted_ URL:

    <link href="{{ '/feed/index.xml' | to_absurl }}" rel="self" />

When you _generate_ your site, the result is:

    <link href="http://www.domain.tld/feed/index.xml" rel="self" />

If you also have `baseurl` set in `_config.yml` then `to_baseurl` will be run before and you'll get:

    <link href="http://www.domain.tld/blog/feed/index.xml" rel="self" />

### method: to\_cdnurl

Prepend a CDN 'virtual versioned' URL for any relative URL resource. Uses up to 4 hosts for multiple download streams at the same time.

In your templates and optionally in your Markdown files use the `to_cdnurl` method to __append__ the `cdn_hosts` parameter value in `_config.yml` to _root-relative_ URLs.

Jekyll's `_config.yml` does not have a `cdn_hosts` parameter, so we need to add our own first. There will likely be other _user specified_ parameters that you will need, so we can namespace all of them in a hash called `app`:

    app:
      release: 100000
      prefix:
      mode: development
      cdn_hosts:
        - xxxxxxxxxxxxx0.cloudfront.net
        - xxxxxxxxxxxxx1.cloudfront.net
        - xxxxxxxxxxxxx2.cloudfront.net
        - xxxxxxxxxxxxx3.cloudfront.net

You will also need a __RewriteRule__ in `.htaccess` to create the 'virtual version' match to the actual asset (assumes `prefix` is 'v'):

    RewriteRule ^v[0-9]{6,6}/(.*)$ /$1 [L]

Note: You do not actually have to rename any files or folders to use 'virtual versioned' URLs. You only need to make sure your 'href' and 'src' attributes point to the virtual version URL. The method does exactly that for you.

Tip: You can _virtual version_ any asset or file just by linking to it using this method! That means PDFs, movies, images, index.html files and web fonts.

Example:

In your `_layouts` -> `default.html` template file you have your __logo__ at the top of the page. You want to serve it from the CDN:

    <object type="image/svg+xml" data="{{ '/img/logo.svg' | to_cdnurl }}">Project Name</object>

When you _generate_ your site, the result is:

    <object type="image/svg+xml" data="//xxxxxxxxxxxxx1.cloudfront.net/v100000/img/logo.svg">Project Name</object>

Depending on the path given to the method, a different CDN host will be chosen using crc32. The final URL is put together like this:

    //xxxxxxxxxxxxxN.cloudfront.net/BASE_URL/PREFIX+RELEASE/img/dog.jpg

Note: `http:` is intentionally left off for protocol anonymous URLs.

See the documentation in the [plugin code](https://github.com/jhauraw/jekyll-url-helper-filter-plugin/blob/master/url-helper-filter.rb) for this method for more information.

### Methods that begin with sub\_

Methods that begin with `sub_` do single or global regex string replacement on the input.

You should use these methods sparingly if at all as it is better to use a `to_` method on the input on a per instance basis.

### methods: sub\_baseurl and sub\_absurl

See the `to_baseurl` and `to_absurl` for basic workings of these methods. The only difference is that the `sub_` methods search through a whole block of input rather than just a short string, meaning multiple instances of _relative_ URLs can be effected.

Example:

One place the `sub_absurl` method is very handy is in your feed xml file. Use it on the `post.content` to make any _relative_ URLs _absolute_.

    post.content (in Markdown):

      Lorem ipsum dolor sit amet, consectetur adipisicing. ![Dog Image](/img/dog.jpg)

    feed.xml (entries loop):

    {% for post in site.posts %}
      <entry>
        ...
        <content type="html">{{ post.content | sub_absurl | xml_escape }}</content>
      </entry>
    {% endfor %}

When you _generate_ your site, the result is:

    <entry>
      ...
      <content type="html">
        Lorem ipsum dolor sit amet, consectetur adipisicing. &lt;a href=&quot;http://domain.tld/img/dog.jpg&quot;&gt;
      </content>
    </entry>

### method: sub\_imgurl

Append image alternate size to images generated in templates such as sidebars and post asides.

This is useful for when you need to show a thumbnail or smaller version of an image such as in a __Post Archive__.

Q: Why not just hardcode the size in the template file?

A: You could. However, if you are specifying a feature image for a post in the `YAML Front Matter` you will be using something like `page.image.src` in your sidebar template and so you can't insert the thumbnail size without using a regex. So moving that task into the plugin was deemed more __DRY__.

Arguments:

  - __size__ -  The string representing the image size. Recommended format is \<width\>x\<height\> measured in pixels.
  - __hires__ - Boolean. Append @2x to the image name if true. For use with retina logic.

Example:

    <img src="{{ '/img/dog.jpg' | sub_imgurl }}" />

When you _generate_ your site, the result is:

    <img src="/img/dog_150x150.jpg" />

Specify __size__:

    <img src="{{ '/img/dog.jpg' | sub_imgurl: '300x300' }}" />

Result:

    <img src="/img/dog_300x300.jpg" />

Specify __size__ and __hires__:

    <img src="{{ '/img/dog.jpg' | sub_imgurl: '300x300', true }}" />

Result:

    <img src="/img/dog_300x300@2x.jpg" />

### method: sanitize\_str

Sanitize a string for use in URLs. Specifically, when you want to link to a __category__ or __tag__ name in your templates.

Category or Tag names with spaces, capital letters, etc. will break your URLs. This method formats them correctly.

Example:

    page.category: Web Development

    <a href="/{{ page.category | sanitize_str }}/">...</a>

When you _generate_ your site, the result is:

    <a href="/web-development/">...</a>

## Summary

Abstracting and normalizing the URL structure of your website is a process that will pay dividends down the road.

Bootstrapers and template developers will especially benefit from this technique when distributing their work to different users who may have preferences of their own in terms of `baseurl` directory name and string names in categories and tags.

## Author

Jhaura Wachsman [website](http://jhaurawachsman.com), [@jhaurawachsman](http://twitter.com/jhaurawachsman)

## CHANGELOG

See [CHANGELOG](https://github.com/jhauraw/jekyll-url-helper-filter-plugin/blob/master/CHANGELOG).

## LICENSE

MIT. See [LICENSE](https://github.com/jhauraw/jekyll-url-helper-filter-plugin/blob/master/LICENSE).

(c) Copyright 2013 Jhaura Wachsman.


CHANGELOG
1.0.0 / 2013-05-24

  - Initial public release.

url-helper-filter.rb
# Jekyll URL Helper Filter Plugin with CDN Support
# Version: 1.0.0
# Release Date: 2013-05-24
# License: MIT
# (c) Copyright Jhaura Wachsman, http://jhaurawachsman.com
module Jekyll
  module Filters

    # See the README for usage examples.

    # The get_ methods grabs/returns a value from _config.yml.

    # Internal: Get the 'url' variable value from _config.yml.
    def get_url
      url = @context.registers[:site].config['url']
    end

    # Internal: Get the 'baseurl' variable value from _config.yml.
    def get_baseurl
      baseurl = @context.registers[:site].config['baseurl']
    end

    # The to_ methods transform a root-relative path into the path
    # type of the method name: to_<type>url.

    # Public: Append the 'baseurl' variable to 'input'.
    def to_baseurl(input)

      # baseurl << input # wouldn't work, created a huge concat chain
      input = "#{get_baseurl}#{input}"
    end

    # Public: Append the 'url' variable to 'input'.
    def to_absurl(input)

      # url + baseurl << input
      input = "#{get_url}#{get_baseurl}#{input}"
    end

    # Public: Append a CDN HTTP host name to a root-relative URL.
    #
    # Requires: Zlib
    #
    # cdn_hosts - An array of up to 4 hosts, ordered from 0 to 3.
    #             Order matters. Pos 0 is host 0, Pos 1 is host 1,
    #             etc. Use Cloudfront hosts if you need SSL, or
    #             CNAMEs otherwise.
    #
    #
    # release   - 6 digit integer. Default is today's date with
    #             two-digit century, e.g., 2025/01/15, release:
    #             250115.
    #
    #             The current site version for use in 'virtual
    #             versioned' URLs and an Origin Pull CDN setup
    #             such as CloudFront. This method does not
    #             require one to use 'invalidate' on every site
    #             update and syncs all assets/files on your web
    #             server with CDN URLs automatically with
    #             Cloudfront, so you don't need to use S3.
    #
    #             Only increment if EXISTING assets need to be
    #             refreshed. Adding new assets to the current
    #             cache does not require a new release number.
    #
    # prefix    - Release id is prefixed with this string. Useful
    #             for catching in an .htaccess RewriteRule.
    #             Default is 'v'.
    #
    # mode      - (development|production). The current working
    #             state. Use 'development' when working locally,
    #             offline or when you want to save bandwidth and
    #             CDN costs. Use 'production' when shipping or
    #             making your code live.
    #
    # Note: Store 'cdn_hosts', 'release', 'prefix' and 'mode' in
    # a user defined hash named 'app' _config.yml. See the README
    # for instructions on how to setup your _config.yml.
    #
    # Note: Using crc32 method to return consistent cross-machine
    # numeric value of string; input.hash sucks and returns diff
    # value for same string, even on same machine.
    #
    # Note: Using value of @input string to match with same CDN
    # Host upon successive site regenerations to maintain caching
    # across builds.
    #
    # Note: You will also need a RewriteRule in .htaccess to
    # create the 'virtual version' match to the actual asset
    # (assumes `prefix` is 'v'):
    #
    # RewriteRule ^v[0-9]{6,6}/(.*)$ /$1 [L]
    #
    # Note: You do not actually have to rename any files or
    # folders to use 'virtual versioned' URLs. You only need to
    # make sure your 'href' and 'src' attributes point to the
    # virtual version URL. The method does exactly that for you.
    #
    # Returns: //xxxxxxxxxxxxxN.cloudfront.net/BASE_URL/PREFIX+RELEASE/img/dog.jpg
    def to_cdnurl(input)

      require 'zlib'

      cdn_hosts = @context.registers[:site].config['app']['cdn_hosts']

      cdn_num = cdn_hosts.length
      hash = Zlib::crc32(input)
      cdn_sub = hash % cdn_num
      cdn_host = cdn_hosts[cdn_sub]

      release = @context.registers[:site].config['app']['release']

      if !release
        release = @context.registers[:site].time.strftime('%y%m%d')
      end

      prefix = @context.registers[:site].config['app']['prefix']

      if !prefix
        prefix = 'v'
      end

      # Debugging
      # puts "\nInput: #{input}\nCDN Host: #{cdn_host}\nCDN Sub: #{cdn_sub}\nCDN Num: #{cdn_num}\nRelease: #{release}\nHash: #{hash}\n//#{cdn_host}#{get_baseurl}/#{prefix}#{release}#{input}\n"

      # If developing locally, point URLs locally, instead of CDN
      if @context.registers[:site].config['app']['mode'] == 'development'
        input = to_baseurl(input)
      else
        input = "//#{cdn_host}#{get_baseurl}/#{prefix}#{release}#{input}"
      end
    end

    # Public: Regex globally append the 'baseurl' variable to 'input'.
    def sub_baseurl(input)
      input.gsub(/(href|src)="\//, "\\1=\"#{get_baseurl}/")
    end

    # Public: Regex globally append the 'url' variable to 'input'.
    def sub_absurl(input)

      # Step 1: Prepend http: protocol to // URLs (protocol anonymous).
      #
      # Step 2: Prepend @url and @baseurl to Root-Relative URLs
      # in templates to make an Absolute URL. Looks for ="/ in
      # href or src tags, not directly followed by a / (look ahead)
      input.gsub(/(href|src)="\/\//, "\\1=\"http://").gsub(/(href|src)="\/(?!\/)/, "\\1=\"#{get_url}#{get_baseurl}/")
    end

    # Public: Regex append image thumbnail sizes and optionally
    # @2x to 'input'.
    def sub_imgurl(input, size = '150x150', hires = nil)

      if hires === true
        hires = "@2x"
      end

      input.sub(/\.(jpg|png|gif)/, "_#{size}#{hires}.\\1")
    end

    # Public: Sanitize a string for use in a URL.
    def sanitize_str(input)
      input.gsub(/[^a-z0-9 -]+/, '').gsub(/\s/, '-').gsub(/-{2,}/, '-').downcase
    end
  end
end

Liquid::Template.register_filter(Jekyll::Filters)
